mailer/engine.py

```diff
@@ -112,6 +113,25 @@ def release_lock(lock):
     logging.debug("released.")
 
 
+def handle_backend_exception(connection, message, err):
+    """ simple error handler that marks messages failing due to SMTP
+    issues as deferred.  Only need to return the action type """
+    error_type = type(err)
+    action = None
+    if isinstance(error_type, (socket_error, smtplib.SMTPSenderRefused,
+        smtplib.SMTPRecipientsRefused, smtplib.SMTPDataError,
+        smtplib.SMTPAuthenticationError):
+
+        message.defer()
+        logging.info("message deferred due to failure: %s" % err)
+        MessageLog.objects.log(message, RESULT_FAILURE, log_message=str(err))
+        action = 'deferred'
+        # Kill the connection, in case the connection itself has an error.
+        connection = None
```

Comment:
`connection = None` does not have any effect on the outer world any more. I think that's by mistake?

Reply:
Honestly no clue anymore.  This pull request was made almost 2 years ago.  Since then I've been using my fork to deal with my needs.  The original code killed off connection so I followed suit.

Reply:
Prior to your last commit, `connection` was returned from the handler and effected on the world outside — what you had was good! This was the working skeleton you had:

```
def handle_backend_exception(connection, message, err):
    action = ........
    if ..........:
        connection = None  # triggers re-connecting outside
    return connection, action

def send_all():
    [..]
        for message in prioritize():
            try:
                if connection is None:
                    connection = get_connection(backend=EMAIL_BACKEND)
                [..]
                email.send()
                [..]
            except Exception as err:
                connection, action = ERROR_HANDLER(connection, message, err)
                counts[action] += 1
```